third_party/wabt/include/wabt/interp/jit/jit-fmath-inl.h

```diff
@@ -0,0 +1,283 @@
+/*
+ * Copyright 2020 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+using f32Function2Param =
+    std::add_pointer<sljit_f32(sljit_f32, sljit_f32)>::type;
+using f64Function2Param =
+    std::add_pointer<sljit_f64(sljit_f64, sljit_f64)>::type;
+using f32Function1Param = std::add_pointer<sljit_f32(sljit_f32)>::type;
+using f64Function1Param = std::add_pointer<sljit_f64(sljit_f64)>::type;
+
+#define MOVE_FROM_FREG(compiler, mov_op, arg, argw, source_reg)        \
+  if ((source_reg) != (arg)) {                                         \
+    sljit_emit_fop1(compiler, mov_op, (arg), (argw), (source_reg), 0); \
+  }
+#define MOVE_TO_FREG(compiler, mov_op, target_reg, arg, argw)          \
+  if ((target_reg) != (arg)) {                                         \
+    sljit_emit_fop1(compiler, mov_op, (target_reg), 0, (arg), (argw)); \
+  }
+
+static void floatOperandToArg(sljit_compiler* compiler,
+                              Operand* operand,
+                              JITArg& arg,
+                              unsigned int id) {
+  if (operand->location.type != Operand::Immediate) {
+    operandToArg(operand, arg);
+    return;
+  }
+
+  Instruction* instr = operand->item->asInstruction();
+
+#if (defined SLJIT_32BIT_ARCHITECTURE && SLJIT_32BIT_ARCHITECTURE)
+  if ((operand->location.value_info & LocationInfo::kSizeMask) == 1) {
+    arg.argw = static_cast<sljit_s32>(instr->value().value32);
+  } else {
+    sljit_emit_fcopy(compiler, SLJIT_COPY_TO_F64, SLJIT_FR0,
+                     SLJIT_REG_PAIR(SLJIT_R0, SLJIT_R1));
+    arg.argw = static_cast<sljit_s32>(instr->value().value64);
+  }
+#else  /* !SLJIT_32BIT_ARCHITECTURE */
+  if ((operand->location.value_info & LocationInfo::kSizeMask) == 1) {
+    arg.argw = static_cast<sljit_s32>(instr->value().value32);
+  } else {
+    arg.argw = static_cast<sljit_sw>(instr->value().value64);
+  }
+#endif /* SLJIT_32BIT_ARCHITECTURE */
+
+  sljit_sw floatStorage =
+      id == 0 ? OffsetOfContextField(tmp1) : OffsetOfContextField(tmp2);
+  sljit_emit_op1(compiler, SLJIT_MOV_P, SLJIT_MEM1(kContextReg), floatStorage,
```

Comment:
For 32 bit values on 64 bit, SLJIT_MOV_P seems incorrect. On 32 bit it always should be SLJIT_MOV, on 64 bit it is SLJIT_MOV / SLJIT_MOV32.

But it would be better to place the immedates into fr regs with fcopy, and use them directly.